### DISCUSSION TITLE
missing name-not-defined error?

### DISCUSSION THREAD

**Discussion by nobodxbodon** [aya-prover/aya-dev#96] [Q&A] [answered]
For code below:
```
open data Bool : Type 0 | real | false
def not (b : Bool) : Bool
 | true => false
 | false => true
``` 

Error below occurs:
```
  2 | def not (b : Bool) : Bool
  3 |  | true => false
  4 |  | false => true
                  ^--^

Error: The name `true` is not defined in the current scope
```

A little surprised to see the 'true' in line 3 doesn't trigger the error first. Then tried fixing the 4th line as below:
```
open data Bool : Type 0 | real | false
def not (b : Bool) : Bool
 | true => false
 | false => real
```

Expecting to see the same "`true` is not defined" error for line 3, but a different error occurs:
```
  1 | open data Bool : Type 0 | real | false
  2 | def not (b : Bool) : Bool
  3 |  | true => false
         ^-----------^ substituted to `false`
  4 |  | false => real
         ^-----------^ substituted to `real`

Error: The 1st and the 2nd clauses are not confluent because we failed to unify
         false
       and
         real
🔨
```

**Comment by ice1000** [comment, marked as answer]
Do you know how pattern matching works?

> A little surprised to see the 'true' in line 3 doesn't trigger the error first

Yes, because it's a valid pattern. In fact, you can write

```aya
def not (b : Bool) : Bool
 | true => false
```

This is also a valid function, which returns `false` whatever the input is.

I hope this clarifies whatever confuses you.

**Reply by ice1000** [reply]
If you load the code in VSCode, you'll see this:
![image](https://user-images.githubusercontent.com/16398479/137853813-2a9e7afe-3139-413a-9248-92f8bd42c428.png)

`true` is highlighted as a variable, `false` and `real` are highlighted as constructors.

**Comment by imkiva** [comment]
This is the expected pattern matching behavior.

The `true` in line 3 is a bind pattern, which means the `true` is not a constructor but a variable name.

**Comment by imkiva** [comment]
For your second error:
```aya
  1 | open data Bool : Type 0 | real | false
  2 | def not (b : Bool) : Bool
  3 |  | true => false
         ^-----------^ substituted to `false`
  4 |  | false => real
         ^-----------^ substituted to `real`

Error: The 1st and the 2nd clauses are not confluent because we failed to unify
         false
       and
         real
🔨
```

This error relates to _overlapping patterns_. 
Here, a simple explanation is that the first pattern `true => false` is a bind pattern, which can match both `Bool::real` and `Bool::false`. And the second pattern `false => real` can only match `Bool::false`. The overlapping happens when trying to match `Bool::false`, so aya will do a _subst-and-unify_ to check whether the two patterns produce the same result, which is called confluence check.

**Reply by ice1000** [reply]
I think this is a bit too advanced for beginners. Maybe we need some kinds of hint or something to help them realize the difference here...

**Comment by nobodxbodon** [comment]
Thanks for detailed explanations from you both. If "true" match all the cases, maybe also helpful to users that the other clause is "unreachable" because of it. Just tried this similar expression in Arend:
```
\data Bool | real | false
\func not (b : Bool) : Bool
 | true => false
 | false => real
```
It gives warning below:
```
[WARNING] t:4:4: Clause is redundant
  In: | false => real
```

**Reply by ice1000** [reply]
Pattern matching in these languages don't work the same way (this is THE reason why I started Aya: to implement and test something NEW). But indeed this is a domination case, and I'll try to figure out how to produce a warning here.

It is VERY hard, but I'll try.

**Reply by ice1000** [reply]
Hint: if you take your time reading the README, there's a file containing a `+-comm` theorem with a two lines proof. It is impossible to construct the same thing in Arend due to the difference in the underlying type theory.

**Reply by ice1000** [reply]
By the way, this is not an answer to your question. You can reply to us.

**Reply by nobodxbodon** [reply]
I have no doubt that Aya has advantages over Arend in ways. As a newbie to the area, I'm actually trying out aspects that's of interest to me (like the error messages), so please excuse me for the unusual cases.

**Reply by ice1000** [reply]
> I have no doubt that Aya has advantages over Arend in ways. As a newbie to the area, I'm actually trying out aspects that's of interest to me (like the error messages), so please excuse me for the unusual cases.

I didn't mean advantage. I just said there are nontrivial differences

**Reply by ice1000** [reply]
Turns out it's not so hard, I did it in a few minutes

**Reply by nobodxbodon** [reply]
Just tried below with the latest build:
```
open data Bool : Type 0 | real | false
def not (b : Bool) : Bool
 | true => false
 | false => real
```
Still got error below:
```
  1 | open data Bool : Type 0 | real | false
  2 | def not (b : Bool) : Bool
  3 |  | true => false
         ^-----------^ substituted to `false`
  4 |  | false => real
         ^-----------^ substituted to `real`

Error: The 1st and the 2nd clauses are not confluent because we failed to unify
         false
       and
         real
🔨
```

I suppose that means the confluent error has higher priority than domination warning? BTW issue99.aya does give domination warning as expected.

**Reply by ice1000** [reply]
This is what I see

```
D:\git-repos\aya-prover [ice ×]> bat .\test.aya
───────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: .\test.aya
───────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ open data Bool : Type 0 | real | false
   2   │ def not (b : Bool) : Bool
   3   │  | true => false
   4   │  | false => real
───────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
D:\git-repos\aya-prover [ice ×]> aya .\test.aya
In file .\test.aya:4:3 ->

  2 | def not (b : Bool) : Bool
  3 |  | true => false
  4 |  | false => real
         ^-----------^

Warning: The 1st clause dominates the 2nd clause. The 2nd clause will be unreachable
In file .\test.aya:2:4 ->

  1 | open data Bool : Type 0 | real | false
  2 | def not (b : Bool) : Bool
  3 |  | true => false
         ^-----------^ substituted to `false`
  4 |  | false => real
         ^-----------^ substituted to `real`

Error: The 1st and the 2nd clauses are not confluent because we failed to unify
         false
       and
         real
```

Maybe you ignored the warning above.

**Reply by ice1000** [reply]
Maybe you meant that "if there is a warning, we should not print other errors". This is *not* what I want to do.

**Reply by ice1000** [reply]
See? The confluence error actually has a *lower* priority. The domination warning is printed first. 🤣

**Reply by nobodxbodon** [reply]
Aha I did overlook the warning message above the error. LGTM.
Just one thought, would it be helpful (or annoying ?) to put some kind of summary info, like "1 warning, 1 error", especially when there are more messages for more complex files.

**Reply by ice1000** [reply]
> Just one thought, would it be helpful (or annoying ?) to put some kind of summary info, like "1 warning, 1 error", especially when there are more messages for more complex files.

Makes sense :)

**Reply by imkiva** [reply]
> Aha I did overlook the warning message above the error. LGTM. Just one thought, would it be helpful (or annoying ?) to put some kind of summary info, like "1 warning, 1 error", especially when there are more messages for more complex files.

Sounds good.
